### DISCUSSION TITLE
Strange SIZE output for -march=rv32ec_zicsr

### DISCUSSION THREAD

**Discussion by sadkotheguest** [xpack-dev-tools/riscv-none-elf-gcc-xpack#16] [Q&A]
I'm trying to compile my project with last x-pack GCC for CH32V003 MCU.

`set(CPU_FLAGS "-march=rv32ec_zicsr -mabi=ilp32e -fmessage-length=0 -msmall-data-limit=0 -msave-restore -fsigned-char -fno-common")`
and 
`set(LD_FLAGS "-march=rv32ec -mabi=ilp32e -Wl,--gc-sections -Wl,--print-memory-usage -nostartfiles")`

And always got 100% RAM usage.

When I'am switching to old one GCC 8.2 (from Mounriver IDE) with 

`set(CPU_FLAGS "-march=rv32ecxw -mabi=ilp32e -fmessage-length=0 -msmall-data-limit=0 -msave-restore -fsigned-char -fno-common")`
and 
`set(LD_FLAGS "-march=rv32ecxw -mabi=ilp32e -Wl,--gc-sections -Wl,--print-memory-usage -nostartfiles")`

everything looks ok.

---
Could you tell where should I dig to make the SIZE (%-age of RAM) work well?

<img width="648" alt="image_2023-03-30_21-52-09" src="https://user-images.githubusercontent.com/82413098/228938468-d9ee5048-86bd-494f-816d-019c018f4051.png">

<img width="647" alt="image_2023-03-30_21-54-08" src="https://user-images.githubusercontent.com/82413098/228938335-d9fc194e-0423-4c9c-a3a2-db3dc8bb5348.png">

**Comment by ilg-ul** [comment]
I suggest you configure the linker to produce the memory map with both compilers and compare the two files, perhaps you can identify some differences.

Anyway, if you use the regular newlib, 2K of RAM might not be enough for the most recent releases. One possible workaround is to try newlib-nano.

**Reply by sadkotheguest** [reply]
Thanks for your answer.
I'am already using newlib-nano.

GCC-12 is using
`riscv-none-elf/lib/rv32ec/ilp32e\libg_nano.a(lib_a-__atexit.o) (__retarget_lock_acquire_recursive)`

Everything else looks quite similar.

[map_files.zip](https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/files/11115449/map_files.zip)

**Reply by sadkotheguest** [reply]
When I am trying to increase the size of the RAM section (up to 500 MB) in .ld file -> 100% of RAM is used also.

**Reply by ilg-ul** [reply]
From what I see in the map, the stack is right at the end of RAM:

```
.stack          0x0000000020000700      0x100
                0x0000000020000700                PROVIDE (_heap_end = .)
                0x0000000020000700                . = ALIGN (0x4)
                [!provide]                        PROVIDE (_susrstack = .)
                0x0000000020000800                . = (. + __stack_size)
 *fill*         0x0000000020000700      0x100
                0x0000000020000800                PROVIDE (_eusrstack = .)
```

It looks like the .ld file is configured to use the entire memory section?

**Reply by sadkotheguest** [reply]
But when I am using GCC 8.2 (with the same .ld) - everything looks OK.

**Reply by ilg-ul** [reply]
It looks, but the map shows the same behaviour, the stack is right at the end. Perhaps the tool used to report the size behaves differently.

**Reply by sadkotheguest** [reply]
Yep.
So my question is about different behaviour of different versions of SIZE utility. :)

**Reply by ilg-ul** [reply]
I think you can update the .ld file to define the correct stack symbols without actually using that top memory, so `size` should report only up to the heap.

**Reply by sadkotheguest** [reply]
Ok, thanks for your advice.

**Reply by sadkotheguest** [reply]
@ilg-ul 
When I am using CH32V307 with the same .ld script (`-march=rv32imafc -mabi=ilp32f`) & the last one GCC 12.2 - SIZE utility works well.

**Reply by ilg-ul** [reply]
I have no idea why.

Try to run manually size from 8.x on both binaries (8.x and 12.x), and similarly size from 12.x on both binaries, and compare the results. If the same binary results in different values, it is a problem with size, which you should report to GNU.

**Reply by sadkotheguest** [reply]
Great idea, thanks.

**Comment by sadkotheguest** [comment]
I found that the subj is the output of the LD with `--print-memory-usage` directive.

```
.stack ORIGIN(RAM) + LENGTH(RAM) - __stack_size :
    {
        . = ALIGN(4);
        PROVIDE(_susrstack = . );
        . = . + __stack_size;
        PROVIDE( _eusrstack = .);
    } >RAM 
```

and

```
.stack :
    {
        . = ALIGN(4);
        PROVIDE(_susrstack = . );
        . = . + __stack_size;
        PROVIDE( _eusrstack = .);
    } >RAM 
```

outputs are different :(

![image_2023-04-01_15-59-46](https://user-images.githubusercontent.com/82413098/229312316-8d08df93-94fe-4ca0-981a-753b9363289c.png)

**Reply by ilg-ul** [reply]
Please note that removing the expression moves the stack right after bss, leaving no space for the heap.

I do not recommend such a change.

**Reply by ilg-ul** [reply]
I also suggest to use the gnu utility size, instead of the linker with that option.

**Reply by sadkotheguest** [reply]
Thanks for your advices.
